### ISSUE TITLE
+ to benchmarks

### ISSUE THREAD

**Issue by tomByrer** [yisar/fre#271]
Would be helpful to compare performance:
https://github.com/krausest/js-framework-benchmark
I'm guessing a bit better than Inferno?

[relevant discussion](https://github.com/krausest/js-framework-benchmark/issues/650)

**Comment by yisar** [comment]
@tomByrer This issue is hard to measure. 

What time slicing improves is response performance, which can make your page no longer blocked.

However, the total time of page rendering is longer than that of synchronous rendering (such as Inferno).

So it's unfair for fre to just measure the rendering time.

Can you understand me?

**Comment by yisar** [comment]
![微信图片_20210606191240](https://user-images.githubusercontent.com/12951461/120924081-7adc8100-c704-11eb-9fe7-fb936218fa35.png)

I refactor the algorithm again, and then used a new technique called `centralized paints`

This is a technology similar to `DocumentFragment`. It operates DOM in memory and finally paints to the screen at final time.

Now fre is super fast, even faster than vanilla js.

**Comment by tomByrer** [comment]
Oh wow, cheers!

**Comment by yisar** [comment]
cc @ryansolid , Hey my friend, I saw your reply on twitter. My country forbids me to login twitter. I want to reply to you here.

This optimization is also called `Offscreen rendering`. Fre operates DOM in memory and paint them to the screen once for the last time.

https://github.com/seedante/iOS-Note/wiki/Mastering-Offscreen-Render

I haven't used the `DocumentFragment` API,  just that their ultimate behavioral results are similar.

I generate an effect list in reverse order in the reconcile phase, and insert them into the screen in reverse order in the commit phase.

This is a pure data structure and algorithm optimization.

Of course, Fiber structure and two-phase traversal do provide convenience, but other singlepass frameworks (such as Vue or preact) are not completely impossible to implement this algorithm.

**Comment by ryansolid** [comment]
To be clear I don't doubt the algorithmic/scheduling improvement. Just when people say faster than vanillajs, some clarification is needed as vanillajs(hand crafted and optimized for the specific solution) by its very nature should be the fastest. So generally I treat that claim with skepticism.

**Comment by yisar** [comment]
> its very nature should be the fastest. So generally I treat that claim with skepticism.

The runtime algorithm with off-screen rendering is faster than vanillajs without off-screen optimization. Maybe it’s more appropriate to say that?

**Comment by yisar** [comment]
@ryansolid 
![微信图片_20210607231854](https://user-images.githubusercontent.com/12951461/121042971-c8311f00-c7e6-11eb-9b12-faf2fa05fea7.png)
![微信图片_20210607231906](https://user-images.githubusercontent.com/12951461/121042977-c8c9b580-c7e6-11eb-87b4-3dc38119aaba.png)

In order to verify the results, I tested again. I was an i7 7700k windows computer, using an M6000 graphics card.

The actual test of fre off-screen rendering has been faster than vanillajs, while solidjs and vanillajs are almost the same.

I tested on another low configuration computer and found that the free off-screen rendering still can be close to vanillajs , which is much better than other vdom frameworks.

**Comment by ryansolid** [comment]
Where/what is this test? When you say offscreen rendering are you also including the cost of applying it on screen. If JS Framework is no good to show this off.. maybe like UIBench(https://github.com/localvoid/uibench)?
Its basically tailored fir comparing VDOM reconcilers.

**Comment by yisar** [comment]
@mindplay-dk  As last time, this test seems unable to accurately test the framework with time slicing. Almost all frameworks use microtask.

**Comment by yisar** [comment]
@mindplay-dk I played the benchmark this afternoon today, but it seems to be slower than expected. 
Next, I will make some adjustments for the test.
At present, it seems that we should do diff props in advance, not here: https://github.com/yisar/fre/blob/master/src/dom.ts#L4
Wait for me.

**Comment by LXSMNSYC** [comment]
There's a lot of improvements to be made:

- https://github.com/yisar/fre/blob/master/src/dom.ts#L9
prefix and suffix spread is usually slower than a single Object.assign call (around <100% slower, specially on chrome)

- https://github.com/yisar/fre/blob/master/src/commit.ts#L30
- https://github.com/yisar/fre/blob/master/src/commit.ts#L40
- https://github.com/yisar/fre/blob/master/src/schedule.ts#L19
- https://github.com/yisar/fre/blob/master/src/reconcile.ts#L229-L230
`forEach` is normally slower than for-increment


- https://github.com/yisar/fre/blob/master/src/schedule.ts#L39
queue shifts is usually slow

- https://github.com/yisar/fre/blob/master/src/h.ts#L20
concat is slower than spread-on-empty-array (70% slower, Chrome)

- https://github.com/yisar/fre/blob/master/src/h.ts#L8
Can be simplified, filter is usually slower than double slice call (60% on 1000+ elements, Chrome)

Still trying to find some core issues. To summarize, most FP methods of arrays are generally slow due to spec-related implementation. It would be better if you can implement/simplify your own array transformations just to provide faster paths.

**Comment by yisar** [comment]
@LXSMNSYC Thank you. You're right, but the most important thing is to implement `fre.memo` This is more important for that benchmark use case. 
This is also the reason why fre slows down.
You said that the optimization of codes is also very reasonable. I will modify them slowly. Thank you!

**Comment by LXSMNSYC** [comment]
Yes. That's one of the issues with Vue before up until 3.2 when they decided to introduce v-memo. If you can implement that then that would provide a lot of render skip process.

**Comment by aleclarson** [comment]
Is this a perf regression, or does Fre really use that much more memory?

![image](https://user-images.githubusercontent.com/1925840/151694973-11e34f38-db11-4e70-8052-554bacb6b79b.png)

**Comment by yisar** [comment]
No, the current benchmark performance metrics for fre are mainly due to the fact that the `fre.memo` API is not implemented, resulting in thousands of duplicate component updates.
https://github.com/krausest/js-framework-benchmark/blob/master/frameworks/keyed/react-hooks/src/main.jsx#L67
